u8timeseries/models/RNN_model.py

```diff
@@ -0,0 +1,469 @@
+
+            self._save_model(False, _get_checkpoint_folder(self.work_dir, self.model_name), epoch)
+
+            if epoch % self.nr_epochs_val_period == 0:
+                training_loss = (total_loss + total_loss_diff) / (batch_idx + 1)  # TODO: do not use batch_idx
+                if val_loader is not None:
+                    validation_loss = self._evaluate_validation_loss(val_loader)
+                    if tb_writer is not None:
+                        tb_writer.add_scalar("validation/loss_total", validation_loss, epoch)
+
+                    if validation_loss < best_loss:
+                        best_loss = validation_loss
+                        self._save_model(True, _get_checkpoint_folder(self.work_dir, self.model_name), epoch)
+
+                    if verbose:
+                        print("Training loss: {:.4f}, validation loss: {:.4f}".
+                              format(training_loss, validation_loss), end="\r")
+                elif verbose:
+                    print("Training loss: {:.4f}".format(training_loss), end="\r")
+
+    def _evaluate_validation_loss(self, val_loader: DataLoader):
+        total_loss = 0
+        total_loss_of_diff = 0
+        self.model.eval()
+        with torch.no_grad():
+            for batch_idx, (data, target) in enumerate(val_loader):
+                data, target = data.to(self.device), target.to(self.device)  # TODO: needed?
+                output = self.model(data)
+                loss = self.criterion(output, target)
+                if self.output_length == 1:
+                    loss_of_diff = self.criterion(output[1:] - output[:-1], target[1:] - target[:-1])
+                else:
+                    loss_of_diff = self.criterion(output[:, 1:] - output[:, :-1], target[:, 1:] - target[:, :-1])
+                total_loss += loss.item()
+                total_loss_of_diff += loss_of_diff.item()
+
+        validation_loss = (total_loss + total_loss_of_diff) / (batch_idx + 1)
+        return validation_loss
+
+    def _save_model(self,
+                    is_best: bool,
+                    folder: str,
+                    epoch: int):
+        """
+        Saves the whole RNNModel object to a file
+        TODO: shall we try to optimize going through torch.save, which uses uses zip?
+
+        :param is_best: whether the model we're currently saving is the best (on validation set)
+        :param folder:
+        :param epoch:
+        :return:
+        """
+
+        checklist = glob(os.path.join(folder, "checkpoint_*"))
+        checklist = sorted(checklist, key=lambda x: float(re.findall('(\d+)', x)[-1]))
+        filename = 'checkpoint_{0}.pth.tar'.format(epoch)
+        os.makedirs(folder, exist_ok=True)
+        filename = os.path.join(folder, filename)
+
+        with open(filename, 'wb') as f:
+            pickle.dump(self, f)
+
+        if len(checklist) >= 5:
+            # remove older files
+            for chkpt in checklist[:-4]:
```

Comment:
isn't it basically checklist[0]?

Reply:
That's `checklist[0]` only when `len(checklist) == 5` but not when `len(checklist) > 5`